crates/uv-auth/src/keyring.rs

```diff
@@ -103,37 +121,74 @@ impl KeyringProvider {
+            // TODO(zanieb): This was done when we added `--mode creds` support for parity with the
+            // existing behavior, but it might be a better UX to hide this on success? It also
+            // might be problematic that we're not streaming it. We could change this given some
+            // user feedback.
+            std::io::stderr().write_all(&output.stderr).ok();
+
             // On success, parse the newline terminated password
-            String::from_utf8(output.stdout)
+            let output = String::from_utf8(output.stdout)
                 .inspect_err(|err| warn!("Failed to parse response from `keyring` command: {err}"))
-                .ok()
-                .map(|password| password.trim_end().to_string())
+                .ok();
+
+            if let Some(username) = username {
+                // We're only expecting a password
+                output.map(|password| (username.to_string(), password.trim_end().to_string()))
+            } else {
+                // We're expecting a username and password
+                output.and_then(|output| {
+                    let mut lines = output.lines();
+                    lines.next().and_then(|username| {
+                        lines
+                            .next()
+                            .map(|password| (username.to_string(), password.to_string()))
+                    })
+                })
+            }
         } else {
             // On failure, no password was available
+            let stderr = std::str::from_utf8(&output.stderr).ok()?;
+            if stderr.contains("unrecognized arguments: --mode") {
+                // N.B. We do not show the `service_name` here because we'll show the warning twice
+                //      otherwise, once for the URL and once for the realm.
+                warn_user_once!(
+                    "Attempted to fetch credentials using the `keyring` command, but it does not support `--mode creds`; upgrade to `keyring>=v25.2.1` for support or provide a username"
+                );
+            } else if username.is_none() {
+                // If we captured stderr, display it in case it's helpful to the user
+                std::io::stderr().write_all(&output.stderr).ok();
+            }
             None
         }
     }
 
     #[cfg(test)]
     fn fetch_dummy(
-        store: &std::collections::HashMap<(String, &'static str), &'static str>,
+        store: &Vec<(String, &'static str, &'static str)>,
         service_name: &str,
-        username: &str,
-    ) -> Option<String> {
-        store
-            .get(&(service_name.to_string(), username))
-            .map(|password| (*password).to_string())
+        username: Option<&str>,
+    ) -> Option<(String, String)> {
+        store.iter().find_map(|(service, user, password)| {
+            if service == service_name && username.map(|username| username == *user).unwrap_or(true)
```

Comment:
```suggestion
            if service == service_name && username.map_or(true, |username| username == *user)
```

Reply:
I don't really find `map_or` easier to reason about personally, but happy to change it.

Reply:
Up to you. It signals to me we're returning a `bool` (with a default for `None`) before reading the closure.

Reply:
I really just prefer it because it's shorter though

Reply:
Clippy actually wants `is_none_or` which is even better

Reply:
Definitely better